Portability fixes

Zer0-One opened the pull request:
- CFLAG gnu99 was changed to c99.
- CXXFLAG c++98 was changed to c++11.
- CFLAG -pedantic-errors was added so that non-ISO C now throws errors.
- _XOPEN_SOURCE feature test macro added and set to 600 to expose SUSv3
  and c99 definitions in modules that required them.
- Fixed tests (and bootstrap daemon logging) that were failing due to
  the altered build flags.
- Avoid string suffix misinterpretation; explicit narrowing conversion.
- Misc. additions to .gitignore to make sure build artifacts don't wind
  up in version control.

I didn't try running the tests, but everything seems to build just fine now. If someone could test for me, I'd really appreciate it.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/toktok/c-toxcore/297)
<!-- Reviewable:end -->

iphydf commented:
Run `other/astyle/format-source`.

iphydf commented:
> Warning suppressions were removed

Indeed, and now it no longer compiles. Please add them back.

GrayHatter commented:
Reviewed 1 of 5 files at r1.
Review status: 1 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[.gitignore, line 25 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KY1DJkp9mHq2fIhmBbh:-KY1DJkp9mHq2fIhmBbi:b-z4xpko) ([raw file](https://github.com/toktok/c-toxcore/blob/653492e93059d079d9db47b364bfc2cb848943fa/.gitignore#L25)):*
> ```
> DartConfiguration.tcl
> CTestTestfile.cmake
> *.pc
> ```

these aren't make files, they're generated files

---

*[CMakeLists.txt, line 73 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KY1FDSoDUFj8vXxM0I4:-KY1FDSpIqIg6gECr8tH:b-7on2fl) ([raw file](https://github.com/toktok/c-toxcore/blob/f0f53dbd5565cbedac30efbfd18202d55d028d1e/CMakeLists.txt#L73)):*
> ```CMake
>   # -pedantic only for C, because in C++ we want to allow the GNU/C99 extension
>   # of having a comma at the end of an enumerator list.
>   add_cflag("-pedantic")
> ```

why did you drop `-pedantic`

---

*[testing/misc_tools.c, line 32 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KY1EiAPxT517MUZb8OH:-KY1EiAQIwNR4q933DMG:bsqx7qx) ([raw file](https://github.com/toktok/c-toxcore/blob/653492e93059d079d9db47b364bfc2cb848943fa/testing/misc_tools.c#L32)):*
> ```C
> #include <stdlib.h>
> #include <string.h>
> #include <strings.h>
> ```

what are you including strings.h for?

---

*[testing/tox_sync.c, line 319 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KY1EThCjujySPlv859L:-KY1EThDdpxBC9hWYkLl:bu14znq) ([raw file](https://github.com/toktok/c-toxcore/blob/653492e93059d079d9db47b364bfc2cb848943fa/testing/tox_sync.c#L319)):*
> ```C
>                     memcpy(filepath+strlen(path), dir->d_name, strlen(dir->d_name)+1);
>                     stat(filepath, &statbuf);
>                     if (S_ISREG(statbuf.st_mode)) {
> ```

why is this better?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

Zer0-One commented:
> these aren't make files, they're generated files

I know that, but does it really matter? What heading do you want me to put them under?

> why did you drop -pedantic

Because I added *-pedantic-errors*, which does the same and more.

> what are you including strings.h for?

Because that's where *strncasecmp()* is declared

> why is this better?

Because *d_type* is a GNU-ism, and the point of this PR is to increase portability.


There's a little bit more for me to fix up to get it to build, pls no mergerino por favorino. I'll let you guys know when it's good.

Zer0-One commented:
@GrayHatter good now?

iphydf commented:
Review status: 1 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[CMakeLists.txt, line 60 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIwsgC2SYUCvkYQX5O:-KYIwsgC3LEep5X8lfHF:bjupbon) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/CMakeLists.txt#L60)):*
> ```CMake
> endmacro()
>   
> # Set standard version for compiler
> ```

Add the '.' back here and add them to the comments below.

---

*[CMakeLists.txt, line 68 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIwy9SE2t_UIqGCR05:-KYIwy9SD0Bd77-VetOV:b-ypfdlo) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/CMakeLists.txt#L68)):*
> ```CMake
> 
> # Expose SUSv3 and C99 definitions
> add_definitions(-D_XOPEN_SOURCE=600)
> ```

Can you add this definition to the files that actually need it rather than the command line?

---

*[other/bootstrap_daemon/src/log.c, line 119 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIxBLP0eYajxNqFLEV:-KYIxBLP1nFJXpFJmkMk:buwb5ot) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/other/bootstrap_daemon/src/log.c#L119)):*
> ```C
> 
>         case LOG_BACKEND_STDOUT:
>             fprintf(level_stdout(level), buf);
> ```

This probably doesn't work. Where are the args?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 1 of 6 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[other/bootstrap_daemon/src/log.c, line 119 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIxBLP0eYajxNqFLEV:-KYIxLht1uNC95orHJes:bo7jwjf) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/other/bootstrap_daemon/src/log.c#L119)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

This probably doesn't work. Where are the args?
</blockquote></details>

Ah. I see now. Don't use `fprintf` with a user-specified format string.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

Zer0-One commented:
@nurupo good catch, i didn't include space for the terminating null.

nurupo commented:
@Zer0-One open Reviewable and comment in it.

Zer0-One commented:
Resolved

---

Review status: 1 of 25 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[.gitignore, line 25 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KY1DJkp9mHq2fIhmBbh:-KY_rxUno4ToGUPkVcWs:bzesk5h) ([raw file](https://github.com/toktok/c-toxcore/blob/653492e93059d079d9db47b364bfc2cb848943fa/.gitignore#L25)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

these aren't make files, they're generated files
</blockquote></details>

We'll clean this up in a different PR, where we shove build artifacts into a directory that *isn't* the root, and then just keep that untracked by git.

---

*[CMakeLists.txt, line 73 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KY1FDSoDUFj8vXxM0I4:-KY_sKvKD9oMxHhXWuuT:b-cr2os1) ([raw file](https://github.com/toktok/c-toxcore/blob/f0f53dbd5565cbedac30efbfd18202d55d028d1e/CMakeLists.txt#L73)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

why did you drop `-pedantic`
</blockquote></details>

Because I added -pedantic-errors

---

*[CMakeLists.txt, line 60 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIwsgC2SYUCvkYQX5O:-KY_sAYK9DNJP0ImQdCV:bb7f2mf) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/CMakeLists.txt#L60)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

Add the '.' back here and add them to the comments below.
</blockquote></details>

done

---

*[CMakeLists.txt, line 68 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIwy9SE2t_UIqGCR05:-KY_sBTEKaHEP6wQNH-A:bb7f2mf) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/CMakeLists.txt#L68)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

Can you add this definition to the files that actually need it rather than the command line?
</blockquote></details>

done

---

*[testing/misc_tools.c, line 32 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KY1EiAPxT517MUZb8OH:-KY_sNSTDgGi6fFsLJsH:b-8u350d) ([raw file](https://github.com/toktok/c-toxcore/blob/653492e93059d079d9db47b364bfc2cb848943fa/testing/misc_tools.c#L32)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

what are you including strings.h for?
</blockquote></details>

Because that's where strncasecmp() is declared

---

*[testing/tox_sync.c, line 319 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KY1EThCjujySPlv859L:-KY_sQOSlShI43q-PjcY:b-9ovlq8) ([raw file](https://github.com/toktok/c-toxcore/blob/653492e93059d079d9db47b364bfc2cb848943fa/testing/tox_sync.c#L319)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

why is this better?
</blockquote></details>

Because d_type is a BSD-ism

---

*[testing/tox_sync.c, line 319 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYKjzej3GBjqZCX2hgI:-KYLbMvs79cykSG9wVRv:bd30ifx) ([raw file](https://github.com/toktok/c-toxcore/blob/f18a53caab7a8bc82944aeebb22670e62dd3293c/testing/tox_sync.c#L319)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

So, you allocate a buffer of size X+Y but you write X+Y+1 into it?
</blockquote></details>

good catch

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-:-KY_rqUsVn_b8bd3chVm:bvm7z8d)*
<!-- Sent from Reviewable.io -->

sudden6 commented:
took a quick look, <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> to me

---

Review status: 1 of 25 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-:-KY_vJzIWLMEAKi_LbWQ:b-v069dm)*
<!-- Sent from Reviewable.io -->

iphydf commented:
@Zer0-One I've pushed a commit. Please squash it into yours if you are ok with that change.

iphydf commented:
Review status: 1 of 26 files reviewed at latest revision, 7 unresolved discussions.

---

*[CMakeLists.txt, line 60 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIwsgC2SYUCvkYQX5O:-KYcVRdGBhfYsdAwMMie:b3vzfci) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/CMakeLists.txt#L60)):*
<details><summary><i>Previously, Zer0-One (David Zero) wrote…</i></summary><blockquote>

done
</blockquote></details>

Not really done. Add `.` to the comment below as well.

---

*[other/bootstrap_daemon/src/log.c, line 119 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIxBLP0eYajxNqFLEV:-KYcV_cACKNsQM4vodzB:b-g9yw7x) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/other/bootstrap_daemon/src/log.c#L119)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

> use vsnprintf with as such

use `vsnprintf` with such*
</blockquote></details>

`write_log(level, "hello %s", "world%s");` what happens if instead of `fprintf(level_stdout(level), "%s", buf)`, this said `fprintf(level_stdout(level), buf)` as it did before?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

nurupo commented:
*[other/bootstrap_daemon/src/log.c, line 119 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIxBLP0eYajxNqFLEV:-KYh9sv9LPwLWVHof6ek:b-psl8jg) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/other/bootstrap_daemon/src/log.c#L119)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

`write_log(level, "hello %s", "world%s");` what happens if instead of `fprintf(level_stdout(level), "%s", buf)`, this said `fprintf(level_stdout(level), buf)` as it did before?
</blockquote></details>

@iphydf There is a misunderstanding here. You were commented on zero-one's change that used `fprintf(level_stdout(level), buf)`, while I was trying to understand why was log.c modified in the first place, assuming that it was modified because of some issue with a user-specified format string. I later read IRC log and found that log.c was modified because `vsyslog()` is not portable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

nurupo commented:
*[other/bootstrap_daemon/src/log.c, line 119 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIxBLP0eYajxNqFLEV:-KYhCMhzxNjo8WUt0dxS:b-c2ar1m) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/other/bootstrap_daemon/src/log.c#L119)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

@iphydf There is a misunderstanding here. You were commented on zero-one's change that used `fprintf(level_stdout(level), buf)`, while I was trying to understand why was log.c modified in the first place, assuming that it was modified because of some issue with a user-specified format string. I later read IRC log and found that log.c was modified because `vsyslog()` is not portable.
</blockquote></details>

@zero-one could you move the code which `vsnprintf()`'s `format` and `args` inside the `log_syslog()` function which you have deleted? And also bring back the `log_stdout()` function. I'm trying to keep syslog and stdout log backends separate.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Reviewed 2 of 5 files at r1, 14 of 20 files at r4, 6 of 8 files at r5.
Review status: 23 of 26 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[CMakeLists.txt, line 65 at r5](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYh7X-mmQ_EMOgTHtq_:-KYh7X-mmQ_EMOgTHtqa:b8u06xf) ([raw file](https://github.com/toktok/c-toxcore/blob/5c795f0b4bd0dd863443fb2c00be9941480ad843/CMakeLists.txt#L65)):*
># Error on non-ISO C

Error? It either needs its own option or be put inside the existing

>if(ERROR_ON_WARNING)

Rationale: builds for users should error.

It's somewhat debatable though, as I'm not sure how likely it's to error for a user but not for us. You explicitly specify to use C99 standard and I wouldn't really expect different compiler versions to exhibit different behavior for C99's `-pedantic-errors`.

---

*[CMakeLists.txt, line 74 at r5](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYh8t-0iUj9dydNimnX:-KYh8t-0iUj9dydNimnY:bptmu4w) ([raw file](https://github.com/toktok/c-toxcore/blob/5c795f0b4bd0dd863443fb2c00be9941480ad843/CMakeLists.txt#L74)):*
> ```CMake
>   add_flag("-Weverything")
> 
> # Disable specific warning flags for both C and C++.
> ```

The comment doesn't have proper indentation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Reviewed 1 of 5 files at r1, 1 of 20 files at r4, 2 of 8 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Pushing ahead to 0.1.1, because nothing is happening here, and we need to focus on 0.1.

Zer0-One commented:
squashed

---

Review status: 24 of 26 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[CMakeLists.txt, line 60 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIwsgC2SYUCvkYQX5O:-KZ6aA1hHGcVGnkAQNND:bb7f2mf) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/CMakeLists.txt#L60)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

Not really done. Add `.` to the comment below as well.
</blockquote></details>

done

---

*[CMakeLists.txt, line 65 at r5](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYh7X-mmQ_EMOgTHtq_:-KZ6aErPFGkHHvhFNMSk:b2dbi6w) ([raw file](https://github.com/toktok/c-toxcore/blob/5c795f0b4bd0dd863443fb2c00be9941480ad843/CMakeLists.txt#L65)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

># Error on non-ISO C

Error? It either needs its own option or be put inside the existing

>if(ERROR_ON_WARNING)

Rationale: builds for users should error.

It's somewhat debatable though, as I'm not sure how likely it's to error for a user but not for us. You explicitly specify to use C99 standard and I wouldn't really expect different compiler versions to exhibit different behavior for C99's `-pedantic-errors`.
</blockquote></details>

non-issue, afaik.

---

*[CMakeLists.txt, line 74 at r5](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYh8t-0iUj9dydNimnX:-KZ6c-6fDOnkNx5kCgyx:b-7ql1td) ([raw file](https://github.com/toktok/c-toxcore/blob/5c795f0b4bd0dd863443fb2c00be9941480ad843/CMakeLists.txt#L74)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

The comment doesn't have proper indentation.
</blockquote></details>

fixed

---

*[other/bootstrap_daemon/src/log.c, line 119 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIxBLP0eYajxNqFLEV:-KZ6c2_0VuAqIiltr1Z9:b7uexcj) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/other/bootstrap_daemon/src/log.c#L119)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

@zero-one could you move the code which `vsnprintf()`'s `format` and `args` inside the `log_syslog()` function which you have deleted? And also bring back the `log_stdout()` function. I'm trying to keep syslog and stdout log backends separate.
</blockquote></details>

ugh, fine. done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-:-KZ6a3j5Xe2621IZBU-u:bn5agbn)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Reviewed 4 of 5 files at r6.
Review status: 26 of 27 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[other/bootstrap_daemon/src/log.c, line 119 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIxBLP0eYajxNqFLEV:-KZB0pc4iay0iVbuEqB7:b-ucfknh) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/other/bootstrap_daemon/src/log.c#L119)):*
<details><summary><i>Previously, Zer0-One (David Zero) wrote…</i></summary><blockquote>

ugh, fine. done
</blockquote></details>

Thanks!

---

*[other/bootstrap_daemon/src/log.c, line 88 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KZB-y4TwWocKw0pHhh8:-KZB-y4TwWocKw0pHhh9:bgof1lk) ([raw file](https://github.com/toktok/c-toxcore/blob/52d06a68676bb95f1d792113bbfe9fbf330cfe5f/other/bootstrap_daemon/src/log.c#L88)):*
> ```C
> 
>     va_copy(args2, args);
>     int size = vsnprintf(NULL, 0, format, args2) + 1;
> ```

Check for negative return value that is returned on error: add both an assert and an if condition that would just return silently.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 26 of 27 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[CMakeLists.txt, line 65 at r5](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYh7X-mmQ_EMOgTHtq_:-KZCivnIpbaIyiIMbfYQ:b-w5eip7) ([raw file](https://github.com/toktok/c-toxcore/blob/5c795f0b4bd0dd863443fb2c00be9941480ad843/CMakeLists.txt#L65)):*
<details><summary><i>Previously, Zer0-One (David Zero) wrote…</i></summary><blockquote>

non-issue, afaik.
</blockquote></details>

I don't have much of an opinion on this. If you make it `-pedantic` and then enable `ERROR_ON_WARNING`, it's the same thing. This `-pedantic-errors` flag is basically `-pedantic -Werror=pedantic`. We can keep it like this, and if we ever run into the situation where this flag causes errors on a user's compiler (or jenkins) but not on our travis build, it stops being a non-issue, and we'll change it to `-pedantic` then.

On the other hand: why do you want to default to erroring on non-ISO C? Given that our travis builds already errors on warnings, this seems to add no value.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 26 of 27 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[other/bootstrap_daemon/src/log.c, line 119 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIxBLP0eYajxNqFLEV:-KZCk0f27YUkqS3gMU48:b23521o) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/other/bootstrap_daemon/src/log.c#L119)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Thanks!
</blockquote></details>

I don't really understand this though. You're replacing one line with another line that calls a function containing the former line, ends up being 500% more code.

The part I don't understand is the "keeping backends separate". You still have explicit knowledge of every backend in this function. What is the advantage of having `my_foo(...)` instead of `foo(...)`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 26 of 27 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[other/bootstrap_daemon/src/log.c, line 119 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIxBLP0eYajxNqFLEV:-KZCknvIG1N_7b1U9HJa:b-dvokyb) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/other/bootstrap_daemon/src/log.c#L119)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

I don't really understand this though. You're replacing one line with another line that calls a function containing the former line, ends up being 500% more code.

The part I don't understand is the "keeping backends separate". You still have explicit knowledge of every backend in this function. What is the advantage of having `my_foo(...)` instead of `foo(...)`?
</blockquote></details>

Never mind. I don't care. As long it works, I'm fine with this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

iphydf commented:
What's the state of this?

nurupo commented:
Waiting on https://reviewable.io/reviews/toktok/c-toxcore/297#-KZB-y4TwWocKw0pHhh8:-KZB-y4TwWocKw0pHhh9:bgof1lk to get resolved.

Zer0-One commented:
done

---

Review status: 14 of 27 files reviewed at latest revision, 6 unresolved discussions.

---

*[CMakeLists.txt, line 65 at r5](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYh7X-mmQ_EMOgTHtq_:-KZGhDCRZRAdaHTK0xVV:bqxkkuh) ([raw file](https://github.com/toktok/c-toxcore/blob/5c795f0b4bd0dd863443fb2c00be9941480ad843/CMakeLists.txt#L65)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

I don't have much of an opinion on this. If you make it `-pedantic` and then enable `ERROR_ON_WARNING`, it's the same thing. This `-pedantic-errors` flag is basically `-pedantic -Werror=pedantic`. We can keep it like this, and if we ever run into the situation where this flag causes errors on a user's compiler (or jenkins) but not on our travis build, it stops being a non-issue, and we'll change it to `-pedantic` then.

On the other hand: why do you want to default to erroring on non-ISO C? Given that our travis builds already errors on warnings, this seems to add no value.
</blockquote></details>

It's more immediately noticeable by a dev who has added non-portable code. i.e, they know right after they run `make` instead of after they've submitted the PR and gone through the Travis logs.

---

*[other/bootstrap_daemon/src/log.c, line 88 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KZB-y4TwWocKw0pHhh8:-KZaMEkfLoGzVnNbCOtS:bb7f2mf) ([raw file](https://github.com/toktok/c-toxcore/blob/52d06a68676bb95f1d792113bbfe9fbf330cfe5f/other/bootstrap_daemon/src/log.c#L88)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Check for negative return value that is returned on error: add both an assert and an if condition that would just return silently.
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-:-KZaMFsqIW5sqeILJDJb:bb7f2mf)*
<!-- Sent from Reviewable.io -->

nurupo commented:
*[other/bootstrap_daemon/src/log.c, line 91 at r7](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KZbh2IArkLdAsm9djUI:-KZbh2IBHP849rxLipBh:b7gko8f) ([raw file](https://github.com/toktok/c-toxcore/blob/426fbc63bd0a6d2e6c36634be5962d7f3b544f5f/other/bootstrap_daemon/src/log.c#L91)):*
> ```C
>     va_end(args2);
> 
>     assert(size < 0);
> ```

     You want to assert that size > 0.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

nurupo commented:
*[other/bootstrap_daemon/src/log.c, line 91 at r7](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KZbh2IArkLdAsm9djUI:-KZbiBwu90LeCYqQZWSQ:b-bmm8ni) ([raw file](https://github.com/toktok/c-toxcore/blob/426fbc63bd0a6d2e6c36634be5962d7f3b544f5f/other/bootstrap_daemon/src/log.c#L91)):*
>= 0 actually, i.e. that it's non-negative.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

nurupo commented:
*[other/bootstrap_daemon/src/log.c, line 91 at r7](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KZbh2IArkLdAsm9djUI:-KZbiWi95gB6FZyGVSfl:b-h8dq0g) ([raw file](https://github.com/toktok/c-toxcore/blob/426fbc63bd0a6d2e6c36634be5962d7f3b544f5f/other/bootstrap_daemon/src/log.c#L91)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

>= 0 actually, i.e. that it's non-negative.
</blockquote></details>

\>= actually, i.e. that it's non-negative.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

Zer0-One commented:
rebased onto latest master

---

Review status: 11 of 27 files reviewed at latest revision, 7 unresolved discussions.

---

*[other/bootstrap_daemon/src/log.c, line 91 at r7](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KZbh2IArkLdAsm9djUI:-KZk9kZ54GrmYz_hrWgy:b-x53ve6) ([raw file](https://github.com/toktok/c-toxcore/blob/426fbc63bd0a6d2e6c36634be5962d7f3b544f5f/other/bootstrap_daemon/src/log.c#L91)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

\>= actually, i.e. that it's non-negative.
</blockquote></details>

lol whoops. fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-:-KZk9erQwkYAUDzyYeAU:b-pwta3q)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Reviewed 2 of 5 files at r1, 4 of 20 files at r4, 3 of 8 files at r5, 10 of 13 files at r7, 6 of 6 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Reviewed 1 of 8 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
:lgtm:

iphydf commented:
@nurupo is this done?

nurupo commented:
Reviewed 11 of 13 files at r7, 6 of 6 files at r8.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[other/bootstrap_daemon/src/log.c, line 88 at r8](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-K_7SoYoMfVk1Naz_uH1:-K_7SoYoMfVk1Naz_uH2:bvl26x6) ([raw file](https://github.com/toktok/c-toxcore/blob/0d12a4b325b3e9878f4a7fbbbfb0453f398ec180/other/bootstrap_daemon/src/log.c#L88)):*
> ```C
> 
>     va_copy(args2, args);
>     int size = vsnprintf(NULL, 0, format, args2) + 1;
> ```

I just noticed this `+ 1` at the end. It breaks the checks we do in `assert` and `if` below, since `vsnprintf()` returns a negative when it errors, which might be `-1`. Also, I don't see any need in having `+1` in there technically speaking, as you already account for `vsnprintf()` not counting null byte by having `+1` in `buf[size + 1]` below.

I think this `+1` should be removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

Zer0-One commented:
finished. rebased onto latest master.

---

Review status: 25 of 27 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[other/bootstrap_daemon/src/log.c, line 88 at r8](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-K_7SoYoMfVk1Naz_uH1:-K_H4UNpRqXG4tiPaVvA:b-59pjz6) ([raw file](https://github.com/toktok/c-toxcore/blob/0d12a4b325b3e9878f4a7fbbbfb0453f398ec180/other/bootstrap_daemon/src/log.c#L88)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

I just noticed this `+ 1` at the end. It breaks the checks we do in `assert` and `if` below, since `vsnprintf()` returns a negative when it errors, which might be `-1`. Also, I don't see any need in having `+1` in there technically speaking, as you already account for `vsnprintf()` not counting null byte by having `+1` in `buf[size + 1]` below.

I think this `+1` should be removed.
</blockquote></details>

you're right, removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-:-K_H4XiONtmFI6aaBgM1:bdbtii8)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Reviewed 1 of 2 files at r9.
Review status: 26 of 27 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[other/bootstrap_daemon/src/log.c, line 98 at r9](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-K_HG3Ji9A4KwE8_GI5e:-K_HG3Ji9A4KwE8_GI5f:btirrsw) ([raw file](https://github.com/toktok/c-toxcore/blob/9ed93d85b432572bfacf0d3bc4e975792466027b/other/bootstrap_daemon/src/log.c#L98)):*
> ```C
> 
>     char buf[size + 1];
>     vsnprintf(buf, size, format, args);
> ```

Ok, found another issue. The size you pass to `vsnprintf()` as the second argument is 1 byte short. What about using `sizeof(buf)` as the second argument?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Reviewed 1 of 2 files at r9.
Review status: 26 of 27 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Review status: 26 of 27 files reviewed at latest revision, 3 unresolved discussions.

---

*[other/bootstrap_daemon/src/log.c, line 119 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-KYIxBLP0eYajxNqFLEV:-K_NFfqTgcGix1IvKkXR:b-iylkif) ([raw file](https://github.com/toktok/c-toxcore/blob/a017f20e6aabead751ac2c8b62a67a6d45537e2d/other/bootstrap_daemon/src/log.c#L119)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

Never mind. I don't care. As long it works, I'm fine with this.
</blockquote></details>

@iphydf you need to okay this to clear reviweable

---

*[other/bootstrap_daemon/src/log.c, line 98 at r9](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-K_HG3Ji9A4KwE8_GI5e:-K_NG7GMnoWhz0BwX7Ya:bpf236z) ([raw file](https://github.com/toktok/c-toxcore/blob/9ed93d85b432572bfacf0d3bc4e975792466027b/other/bootstrap_daemon/src/log.c#L98)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Ok, found another issue. The size you pass to `vsnprintf()` as the second argument is 1 byte short. What about using `sizeof(buf)` as the second argument?
</blockquote></details>

or init size like `int size = 1 + vsnprintf(...);`

then change the if to be `<=` and you can simply use size.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Reviewed 1 of 2 files at r9.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Reviewed 3 of 5 files at r1, 4 of 20 files at r4, 4 of 8 files at r5, 10 of 13 files at r7, 4 of 6 files at r8, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 1 of 1 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-:-K_RAbGPiLp9v9pCiQtP:bg7igkv)*
<!-- Sent from Reviewable.io -->

Zer0-One commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[other/bootstrap_daemon/src/log.c, line 98 at r9](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-K_HG3Ji9A4KwE8_GI5e:-K_RCfn0FIyoGizrvd80:bzczsha) ([raw file](https://github.com/toktok/c-toxcore/blob/9ed93d85b432572bfacf0d3bc4e975792466027b/other/bootstrap_daemon/src/log.c#L98)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

or init size like `int size = 1 + vsnprintf(...);`

then change the if to be `<=` and you can simply use size.
</blockquote></details>

fixed, in what i think is the most readable way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[CMakeLists.txt, line 69 at r10](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-K_UW_ZWww_qtTLqpgk6:-K_UW_ZXnnTmn_3fgTOW:b-3e1eav) ([raw file](https://github.com/toktok/c-toxcore/blob/b00e38a21de35ad931b73803c4d9a7dc0e28f6a0/CMakeLists.txt#L69)):*
> ```CMake
> option(WARNINGS "Enable additional compiler warnings" ON)
> if(WARNINGS)
> # Add all warning flags we can.
> ```

Indent comments with 2 spaces just like the rest of the block.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

Zer0-One commented:
Review status: 26 of 27 files reviewed at latest revision, 2 unresolved discussions.

---

*[CMakeLists.txt, line 69 at r10](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-K_UW_ZWww_qtTLqpgk6:-K_WtX43WtQ4UJnFDLwx:bb7f2mf) ([raw file](https://github.com/toktok/c-toxcore/blob/b00e38a21de35ad931b73803c4d9a7dc0e28f6a0/CMakeLists.txt#L69)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

Indent comments with 2 spaces just like the rest of the block.
</blockquote></details>

done

---

*[other/bootstrap_daemon/src/log.c, line 98 at r9](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-K_HG3Ji9A4KwE8_GI5e:-K_WtZh_sVVUdiJf0FRw:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/9ed93d85b432572bfacf0d3bc4e975792466027b/other/bootstrap_daemon/src/log.c#L98)):*
<details><summary><i>Previously, Zer0-One (David Zero) wrote…</i></summary><blockquote>

fixed, in what i think is the most readable way.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297)*
<!-- Sent from Reviewable.io -->

nurupo commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r10, 2 of 2 files at r11.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-:-K_faYx9thkD4NQTOaRa:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 2 of 2 files at r11.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/297#-:-K_hR1oICNptMomD6Y6q:bg7igkv)*
<!-- Sent from Reviewable.io -->
